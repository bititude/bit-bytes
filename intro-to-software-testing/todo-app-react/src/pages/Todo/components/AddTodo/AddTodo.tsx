import { useState } from 'react';
import Spinner from '../../../../components/Spinner/Spinner';
import { useTodo } from '../../hooks/useTodo';

const AddTodo: React.FC = () => {
  const [title, setTitle] = useState('');
  const { isAdding, addTodo, error } = useTodo();

  return (
    <div data-testid='add-todo-form'>
      <textarea
        disabled={isAdding}
        onChange={e => setTitle(e.target.value)}
        value={title}
        className='w-full border-2 border-gray-400 rounded-md p-3 font-bold mb-4 shadow-md'
      ></textarea>
      <div className='flex items-start w-full justify-between'>
        <button
          onClick={() => addTodo({ title, completed: false, id: Date.now().toString() })}
          disabled={isAdding || !title}
          className='bg-green-500 h-[50px] font-bold text-white rounded-md shadow-md hover:scale-105 mb-4 text-base disabled:cursor-not-allowed w-[100px] flex items-center justify-center hover:shadow-lg disabled:bg-gray-400'
        >
          <span className='text-lg'>ADD</span>
          {isAdding ? (
            <span data-testid='spinner' className='text-4xl ml-3'>
              <Spinner />
            </span>
          ) : (
            <span className='text-4xl ml-3 pb-2'>&#43;</span>
          )}
        </button>
      </div>
      {error && error.type == 'ADD' ? <p className='text-red-600'>{error.error}</p> : null}
    </div>
  );
};

export default AddTodo;
