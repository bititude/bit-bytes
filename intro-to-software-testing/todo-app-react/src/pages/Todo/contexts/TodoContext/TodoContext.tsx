import { createContext, ReactNode, useCallback, useEffect, useState } from 'react';
import { Todo } from '../../../../types/Todo';
import {
  createTodo,
  getTodos,
  markTodoAsComplete,
  removeTodo,
} from '../../../../services/api/todo/todo';

type TodoError = {
  type: 'ADD' | 'LOAD' | 'DELETE' | 'UPDATE';
  error: string;
};

type TodoContextType = {
  todos: Todo[];
  isAdding: boolean;
  addTodo: (todo: Todo) => void;
  deleteTodo: (id: string) => void;
  markAsCompleted: (id: string) => void;
  error: TodoError | null;
  isTodosLoading: boolean;
};

export const TodoContext = createContext<TodoContextType | undefined>(undefined);

export const TodoProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [todos, setTodos] = useState<Todo[]>([]);
  const [isAdding, setIsAdding] = useState(false);
  const [error, setError] = useState<TodoError | null>(null);
  const [isTodosLoading, setIsTodosLoading] = useState(false);

  const addTodo = useCallback(async (todo: Todo) => {
    try {
      setIsAdding(true);
      await createTodo(todo);
      setTodos(currentTodos => [todo, ...currentTodos]);
    } catch {
      setError({
        type: 'ADD',
        error: 'something went wrong creating the todo, please try again later',
      });
    } finally {
      setIsAdding(false);
    }
  }, []);

  const deleteTodo = useCallback(async (id: string) => {
    try {
      await removeTodo(id);
      setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
    } catch {
      setError({
        type: 'DELETE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const markAsCompleted = useCallback(async (id: string) => {
    try {
      await markTodoAsComplete(id);
      setTodos(currentTodos =>
        currentTodos.map(todo => (todo.id === id ? { ...todo, completed: true } : todo))
      );
    } catch {
      setError({
        type: 'UPDATE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const loadTodos = useCallback(async () => {
    try {
      setIsTodosLoading(true);
      const data = await getTodos();
      setTodos(data);
    } catch {
      setError({
        type: 'LOAD',
        error: "sorry, we can't fetch your todos right now, please try again later",
      });
    } finally {
      setIsTodosLoading(false);
    }
  }, []);

  useEffect(() => {
    loadTodos();
  }, [loadTodos]);

  return (
    <TodoContext.Provider
      value={{ todos, addTodo, markAsCompleted, deleteTodo, error, isAdding, isTodosLoading }}
    >
      {children}
    </TodoContext.Provider>
  );
};
