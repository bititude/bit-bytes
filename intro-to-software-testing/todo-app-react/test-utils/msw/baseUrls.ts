/// <reference types="vite/client" />

export const apiEndPoint = (path: string) => {
  return import.meta.env.VITE_API_BASE_URL + path;
};
