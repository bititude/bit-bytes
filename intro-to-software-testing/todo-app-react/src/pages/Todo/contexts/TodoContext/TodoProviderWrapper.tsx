import { TodoProvider } from './TodoContext';

export const TodoProviderWrapper: React.FC = ({ children }) => (
  <TodoProvider>{children}</TodoProvider>
);
