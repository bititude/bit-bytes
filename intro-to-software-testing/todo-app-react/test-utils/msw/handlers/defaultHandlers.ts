import { delay, HttpResponse } from 'msw';

export default {
  500: () =>
    HttpResponse.json({ statusCode: 500, message: 'Internal server error' }, { status: 500 }),
  ERR_NETWORK: () => HttpResponse.error(),
  LOADING: async () => {
    await delay('real');
    return HttpResponse.json([]);
  },
  INFINITE_LOADING: async () => {
    await delay('infinite');
  },
};
