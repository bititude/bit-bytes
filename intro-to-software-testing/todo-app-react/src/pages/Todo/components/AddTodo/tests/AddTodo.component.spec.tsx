import { describe, expect, it } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import AddTodo from '../AddTodo';
import { server } from '../../../../../../test-utils/msw/node';
import { apiEndPoint } from '../../../../../../test-utils/msw/baseUrls';
import defaultHandlers from '../../../../../../test-utils/msw/handlers/defaultHandlers';
import { http } from 'msw';
import { TodoProviderWrapper } from '../../../contexts/TodoContext/TodoProviderWrapper';

// Behaviour driven test cases (BDD) - Given (precondition) -> When (action) -> Then (outcome), assert on the outcome

/**
 * Requirements:-
 * 1. when add todo button is clicked,
 * 1.1 textarea should be disabled and be enabled after adding todo
 * 1.2 add button should be disabled if textarea is empty
 * 1.3 add button should be disabled and enabled after adding
 * 1.4 a spinner should be shown and it should dissappear after adding
 * 1.5 an error message should be shown if add todo is failure
 */

// use component name or function name
describe('AddTodo.tsx', () => {
  // helper methods
  const typeTodo = () => {
    const textArea = screen.getByRole('textbox');
    fireEvent.change(textArea, { target: { value: 'My new task' } });
  };

  // when
  describe('when add todo button is clicked', () => {
    // then
    it('textarea should be disabled and be enabled after adding todo', async () => {
      // render the component
      render(<AddTodo />, { wrapper: TodoProviderWrapper });

      // get reference to all the required elements in the ui
      //https://testing-library.com/docs/queries/about#priority
      // https://testing-library.com/docs/react-testing-library/cheatsheet/
      const textArea = screen.getByRole('textbox');
      const button = screen.getByRole('button', { name: /add/i });

      typeTodo();

      // pre assertion
      expect(textArea).not.toBeDisabled();

      // specify msw handler for loading state
      server.use(http.post(apiEndPoint('/todo'), defaultHandlers['LOADING']));

      // perform the action
      fireEvent.click(button);

      // After clicking, the textarea should be disabled while adding
      expect(textArea).toBeDisabled();

      await waitFor(() => {
        expect(textArea).not.toBeDisabled();
      });
    });

    it('button should be disabled and enabled after adding', async () => {
      render(<AddTodo />, { wrapper: TodoProviderWrapper });

      const addButton = screen.getByRole('button', { name: /add/i });

      expect(addButton).toBeDisabled();

      typeTodo();

      expect(addButton).not.toBeDisabled();

      server.use(http.post(apiEndPoint('/todo'), defaultHandlers['LOADING']));

      fireEvent.click(addButton);

      expect(addButton).toBeDisabled();

      await waitFor(() => {
        expect(addButton).toBeDisabled();
      });
    });

    it('a spinner should be shown and it should dissappear after adding', async () => {
      render(<AddTodo />, { wrapper: TodoProviderWrapper });

      const addButton = screen.getByRole('button', { name: /add/i });

      expect(screen.queryByTestId('spinner')).toBe(null);

      typeTodo();

      server.use(http.post(apiEndPoint('/todo'), defaultHandlers['LOADING']));

      fireEvent.click(addButton);

      await waitFor(() => {
        expect(screen.queryByTestId('spinner')).toBeInTheDocument();
      });

      await waitFor(() => {
        expect(screen.queryByTestId('spinner')).not.toBeInTheDocument();
      });
    });

    it('add button should be disabled if textarea is empty', () => {
      render(<AddTodo />, { wrapper: TodoProviderWrapper });

      const addButton = screen.getByRole('button', { name: /add/i });

      expect(addButton).toBeDisabled();

      typeTodo();

      expect(addButton).toBeEnabled();
    });

    it('an error message should be shown if add todo is failure', async () => {
      render(<AddTodo />, { wrapper: TodoProviderWrapper });

      const addButton = screen.getByRole('button', { name: /add/i });

      typeTodo();

      server.use(http.post(apiEndPoint('/todo'), defaultHandlers['500']));

      fireEvent.click(addButton);

      await waitFor(() => {
        expect(
          screen.getByText('something went wrong creating the todo, please try again later')
        ).toBeInTheDocument();
      });
    });
  });

  describe('should match snapshots', () => {
    it("match's default snapshot", () => {
      const { asFragment } = render(<AddTodo />, { wrapper: TodoProviderWrapper });
      expect(asFragment()).toMatchSnapshot();
    });

    it("match's loading state snapshot", () => {
      const { asFragment } = render(<AddTodo />, { wrapper: TodoProviderWrapper });

      const addButton = screen.getByRole('button', { name: /add/i });

      typeTodo();

      server.use(http.post(apiEndPoint('/todo'), defaultHandlers['LOADING']));

      fireEvent.click(addButton);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
