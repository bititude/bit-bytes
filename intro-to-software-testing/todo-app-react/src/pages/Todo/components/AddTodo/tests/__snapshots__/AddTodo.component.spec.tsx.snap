// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddTodo.tsx > should match snapshots > match's default snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="add-todo-form"
  >
    <textarea
      class="w-full border-2 border-gray-400 rounded-md p-3 font-bold mb-4 shadow-md"
    />
    <div
      class="flex items-start w-full justify-between"
    >
      <button
        class="bg-green-500 h-[50px] font-bold text-white rounded-md shadow-md hover:scale-105 mb-4 text-base disabled:cursor-not-allowed w-[100px] flex items-center justify-center hover:shadow-lg disabled:bg-gray-400"
        disabled=""
      >
        <span
          class="text-lg"
        >
          ADD
        </span>
        <span
          class="text-4xl ml-3 pb-2"
        >
          +
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddTodo.tsx > should match snapshots > match's loading state snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="add-todo-form"
  >
    <textarea
      class="w-full border-2 border-gray-400 rounded-md p-3 font-bold mb-4 shadow-md"
      disabled=""
    >
      My new task
    </textarea>
    <div
      class="flex items-start w-full justify-between"
    >
      <button
        class="bg-green-500 h-[50px] font-bold text-white rounded-md shadow-md hover:scale-105 mb-4 text-base disabled:cursor-not-allowed w-[100px] flex items-center justify-center hover:shadow-lg disabled:bg-gray-400"
        disabled=""
      >
        <span
          class="text-lg"
        >
          ADD
        </span>
        <span
          class="text-4xl ml-3"
          data-testid="spinner"
        >
          <svg
            class="mr-3 h-5 w-5 animate-spin text-accent"
            fill="none"
            viewBox="0 0 24 24"
          >
            <circle
              class="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              stroke-width="4"
            />
            <path
              class="opacity-75"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
